### DISCUSSION TITLE
Is there anything like the VS Code Testing API that has been implemented?

### DISCUSSION THREAD

**Discussion by ckipp01** [mfussenegger/nvim-dap#414] [Q&A]
Recently in Metals we added support for [test discovery](https://github.com/scalameta/metals/pull/3277) to support the [VS Code Testing API](https://vscode.trafficmanager.net/api/extension-guides/testing). Under the hood tests are still actually ran via DAP, and I was curious to play around with something like this to easily allow a user to run individuals tests or suites with a more finer granularity then the "test target" or "test file" that Metals currently has. However I was curious if something like this has ever been explored or if someone has already played around with this? Basically something similar to the Packer window that opens up and you'd be able to trigger tests and see them update using ext marks.

<img width="1071" alt="Screenshot 2022-01-17 at 17 28 55" src="https://user-images.githubusercontent.com/13974112/149806879-9aac0bd2-fbe1-4e65-b3fc-aa55d76b0f9e.png">

Do you know if this has been looked into at all?

**Comment by mfussenegger** [comment]
As far as I'm aware there is no unified API or something like that yet.

When I started out with nvim-dap and nvim-jdtls I briefly looked at https://build-server-protocol.github.io/ as it would provide the functionality to discover build targets and to run them and is piggybacking onto LSP & DAP, but unfortunately outside of the scala world it afaik hasn't really seen any adaption (yet).

If it did, I think it would be nice to have a unifying test-discovery plugin for neovim as well (that would utilize nvim-dap and the built-in lsp client)

Other than that there is https://github.com/rcarriga/vim-ultest as a testing framework (with nvim-dap support)


All the extensions so far provide dedicated "test nearest method" functions:

- https://github.com/leoluz/nvim-dap-go
- https://github.com/David-Kunz/jester
- https://github.com/mfussenegger/nvim-dap-python
- https://github.com/mfussenegger/nvim-jdtls


They just search for the nearest test (often via tree-sitter) and then call `dap.run()` with a ad-hoc generated configuration.

`nvim-jdtls` also provides a `pick_test` to pick from all tests within a file. The test results are appended to the dap-repl and it also populates the quickfix list. See https://github.com/mfussenegger/nvim-jdtls/pull/39

**Reply by ckipp01** [reply]
Awesome, thanks a lot for all of this @mfussenegger, super helpful. I'll dig around these and probably come up with some sort of minimal POC. I like the of `vim-ultest` but at least for what I need in `nvim-metals` it can be greatly simplified to just a some lsp calls followed by a hand off to `nvim-dap` so I'll play around with this. I have my plan outlined [here](https://github.com/scalameta/nvim-metals/issues/324) if you're curious. Again, thanks!
